File: python/yacht/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 1
ONES = 2
TWOS = 3
THREES = 4
FOURS = 5
FIVES = 6
SIXES = 7
FULL_HOUSE = 8
FOUR_OF_A_KIND = 9
LITTLE_STRAIGHT = 10
BIG_STRAIGHT = 11
CHOICE = 12


def score(dice, category):
    match category:
        case 1:
            return 50 if dice in [[1] * 5, [2] * 5, [3] * 5, [4] * 5, [5] * 5, [6] * 5] else 0
    
        case 2:
            r = 0
            for n in dice:
                if n == 1:
                    r += 1
            return r
    
        case 3:
            r = 0
            for n in dice:
                if n == 2:
                    r += 2
            return r
    
        case 4:
            r = 0
            for n in dice:
                if n == 3:
                    r += 3
            return r
    
        case 5:
            r = 0
            for n in dice:
                if n == 4:
                    r += 4
            return r
    
        case 6:
            r = 0
            for n in dice:
                if n == 5:
                    r += 5
            return r
    
        case 7:
            r = 0
            for n in dice:
                if n == 6:
                    r += 6
            return r
    
        case 8:
            n1 = None
            n2 = None
            n1_count = 1
            n2_count = 1
    
            for n in dice:
                if n1 is None:
                    n1 = n
                    continue
                if n is n1:
                    n1_count += 1
                    continue

                if n2 is None:
                    n2 = n
                    continue
                if n is n2:
                    n2_count += 1
                    continue

                if n is not n1 and n is not n2:
                    return 0

            return sum(dice) if n1_count == 3 or n2_count == 3 else 0
    
        case 9:
            n1 = None
            n2 = None
            n1_count = 1
            n2_count = 1
    
            for n in dice:
                if n1 is None:
                    n1 = n
                    continue
                if n is n1:
                    n1_count += 1
                    continue

                if n2 is None:
                    n2 = n
                    continue
                if n is n2:
                    n2_count += 1
                    continue

                if n is not n1 and n is not n2:
                    return 0

            if n1_count >= 4:
                return n1 * 4
            if n1_count >= 4:
                return n2 * 4

            return 0
    
        case 10:
            return 30 if sorted(dice) == [1, 2, 3, 4, 5] else 0
    
        case 11:
            return 30 if sorted(dice) == [2, 3, 4, 5, 6] else 0
    
        case 12:
            return sum(dice)

File: fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let dieToNum (die: Die): int =
    match die with
    | One -> 1
    | Two -> 2
    | Three -> 3
    | Four -> 4
    | Five -> 5
    | Six -> 6
    
let doFullHouse (dice: List<Die>): int =
    let mutable one: Option<Die> = None
    let mutable oneCount = 1
    let mutable two: Option<Die> = None
    let mutable twoCount = 1

    let rec check' d =
        match d with
        | [] -> ()
        | h::t ->
            if one = None then one <- Some h
            elif two = None && (Option.get one) <> h then two <- Some h
            elif (Option.get one) = h then oneCount <- oneCount + 1
            elif (Option.get two) = h then twoCount <- twoCount + 1
            check' t

    check' dice

    match oneCount, twoCount with
    | 3, 2 | 2, 3 -> 
        (dieToNum (Option.get one)) * oneCount + (dieToNum (Option.get two)) * twoCount
    | _, _ -> 0

let doFourOfKind (dice: List<Die>): int =
    let mutable one: Option<Die> = None
    let mutable oneCount = 1
    let mutable two: Option<Die> = None
    let mutable twoCount = 1

    let rec check' d =
        match d with
        | [] -> ()
        | h::t ->
            if one = None then one <- Some h
            elif two = None && (Option.get one) <> h then two <- Some h
            elif (Option.get one) = h then oneCount <- oneCount + 1
            elif (Option.get two) = h then twoCount <- twoCount + 1
            check' t

    check' dice

    match oneCount, twoCount with
    | c, _ when c >= 4 -> (dieToNum (Option.get one)) * 4
    | _, c when c >= 4 -> (dieToNum (Option.get two)) * 4
    | _, _ -> 0

let rec doCmpList l1 l2 =
    match l1, l2 with
    | [], [] -> true
    | x::xs, y::ys -> if x <> y then false else doCmpList xs ys
    | _, _ -> false

let score (category: Category) (dice: List<Die>): int = 
    match category with
    | Ones -> List.filter (fun die -> die = One) dice |> List.length
    | Twos -> List.filter (fun die -> die = Two) dice |> List.length |> (*) 2
    | Threes -> List.filter (fun die -> die = Three) dice |> List.length |> (*) 3
    | Fours -> List.filter (fun die -> die = Four) dice |> List.length |> (*) 4
    | Fives -> List.filter (fun die -> die = Five) dice |> List.length |> (*) 5
    | Sixes -> List.filter (fun die -> die = Six) dice |> List.length |> (*) 6
    | FullHouse -> doFullHouse dice
    | FourOfAKind -> doFourOfKind dice
    | LittleStraight -> List.sort dice |> (fun sorted -> if sorted = [One; Two; Three; Four; Five] then 30 else 0)
    | BigStraight -> List.sort dice |> (fun sorted -> if sorted = [Two; Three; Four; Five; Six] then 30 else 0)
    | Choice -> List.map (dieToNum) dice |> List.reduce (+)
    | Yacht ->
        if (doCmpList dice [One; One; One; One; One] ||
            doCmpList dice [Two; Two; Two; Two; Two] ||
            doCmpList dice [Three; Three; Three; Three; Three] ||
            doCmpList dice [Four; Four; Four; Four; Four] ||
            doCmpList dice [Five; Five; Five; Five; Five] ||
            doCmpList dice [Six; Six; Six; Six; Six]) then 50 else 0